feat(websocket): Expanded example to demonstrate the transfer over TLS (IDFGH-11681)

gabsuren opened the pull request.

david-cermak commented on components/esp_websocket_client/examples/target/main/certs/client_csr.pem:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE REQUEST-----
+MIICijCCAXICAQAwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgMClNvbWUtU3RhdGUx
+ITAfBgNVBAoMGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZDCCASIwDQYJKoZIhvcN
+AQEBBQADggEPADCCAQoCggEBAPjFGt/rLI1zAcopnTfNF5NxJpSJ5DJYvTq4Us2G
+sGEftJn+mYHdjRk0JVIiTT4DDxeYlVhg51og1dBL0sJLjZyx99BkJPuIlV0g9VDG
+K/y684mx0H8Mz5SLfXy+XSm/32ycL1ux6YIBroiF5bC+YJf86AJ86JgG80U4EOsj
+GlOwnR6l3681Kso7NYVUzMCufdH8p3xC7yfmgABkvPGcs3wxWIrCvv+oBOOVmKhe
+usubOaOrs2k9VdT7b1e56cZrMWr/ztHe6TAZQ7OcHJI09t3P55XJgOBUSQq0G9JG
+QYvtMn3zSXW23mnFystPv+KV06yeoHhIaItK+i1ZMkNFlBUCAwEAAaAAMA0GCSqG
+SIb3DQEBCwUAA4IBAQCrne3RyK+HPjZ22qXzs/SdAH7OFXrFmCxyiWDarHCkt6yK
+/YllyeaQLym7vP7KT1iJqXsj1R7T+alKPF2BJ2cW8O2MAPWPC4aQ8PCYQpmy9990
+0KdQzfZKvr97o/O8Qm1GSauSXWmz1drbUf1LGKKliFaXmBTmJ+rt9hpKZNROQrpA
+qYAaGWrD26dizB9KQKX5vOghCHULkcjgPddUEbhPMaClw5uC8rhLuGghfs7IbFWB
+oGvKXAScONLZbfu+UIz2i0z138njXxkMBK7EWOYb+oapkrVJp9LlsbxQD0bwkagM
+5o5V+EGSbc+d19Birhc8ut/7BWw9F1lQHHmR6Meb
+-----END CERTIFICATE REQUEST-----
```
We don't have to add cert signing request to git.

david-cermak commented on components/esp_websocket_client/examples/target/main/certs/ca_cert.srl:
```diff
@@ -0,0 +1 @@
+205DA749D4B875743377BED4CE4F21EC0A4294A1
```
same here, suggest removing

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,47 @@ This example can be executed on any ESP32 board, the only required interface is
+
+### Server Certificate Verification
+
+* To skip server certificate verification, enable `ESP_TLS_SKIP_SERVER_CERT_VERIFY=y`
```
I don't think we should suggest skipping server's certificate in our examples

david-cermak commented:
As mentioned in the comments, I'd suggest using server side verification (always), optionally adding client side verification (mutual authentication).

I think for user's perspective it would be useful to see these two additional configs:
1) TLS with server side verification only (using some public ws server if available)
2) Mutual authentication  -- with self signed certificates

gabsuren commented:
@david-cermak thank you for the review. 
I've updated the example with the following changes:

1. The server now always checks the client's certificates `ssl_context.verify_mode = ssl.CERT_REQUIRED`.
2. Both client and server now perform mutual authentication using self-signed certificates.
2. The client bypasses server certificate verification by enabling CONFIG_ESP_TLS_SKIP_SERVER_CERT_VERIFY=y. But, it uses public and private keys, allowing the server to authenticate the client's identity through its certificates

Please let me know if this adjustment looks ok for you

gabsuren commented on components/esp_websocket_client/examples/target/sdkconfig.ci.tls_cert_not_required:
```diff
@@ -0,0 +1,16 @@
+CONFIG_EXAMPLE_CONNECT_IPV6=y
+CONFIG_WS_OVER_TLS_SCHEME=y # this config only needs for python script configuration
+CONFIG_ESP_TLS_SKIP_SERVER_CERT_VERIFY=y
+CONFIG_ESP_TLS_INSECURE=y
```
This setup applies when the server requires client certificates, while the client skips verifying the server's certificates. 
As it was in your initial suggestion` 1. TLS with server side verification only (using some public ws server if available)`

david-cermak replied:
sorry, what I meant by:

> TLS with server side verification only (using some public ws server if available)

is that the client only verifies the server

gabsuren commented:
Updated : 
Modified the example to have two configuration 
* Mutual Authentication: When `CONFIG_WS_OVER_TLS_MUTUAL_AUTH=y` is enabled, it's essential to provide valid certificates for both the server and client.
  This ensures a secure two-way verification process.
* Server-Only Authentication: To perform verification of the server's certificate only (without requiring a client certificate), set `CONFIG_WS_OVER_TLS_SERVER_AUTH=y`.

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,52 @@ This example can be executed on any ESP32 board, the only required interface is
+  This ensures a secure two-way verification process.
+* Server-Only Authentication: To perform verification of the server's certificate only (without requiring a client certificate), set `CONFIG_WS_OVER_TLS_SERVER_AUTH=y`.
+  This method skips client certificate verification.
+* Example below demonstrates how to generate a new certificate for the server and client using the OpenSSL command line tool
```
need to mention that these are *self-signed* certs

...how to generate self-signed certificates...

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,52 @@ This example can be executed on any ESP32 board, the only required interface is
+* Example below demonstrates how to generate a new certificate for the server and client using the OpenSSL command line tool
+
+### Generating Certificates with OpenSSL
+* The example below outlines the process for creating new certificates for both the server and client using OpenSSL, a widely-used command line tool for implementing SSL and TLS protocols:
```
>  implementing SSL and TLS protocols:

I think that SSL and TLS are just different names for the same thing

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,52 @@ This example can be executed on any ESP32 board, the only required interface is
+Generate a Certificate Signing Request (CSR) for the Server
+openssl req -new -key server_key.pem -out server_csr.pem
+
+Sign the Server's CSR with the CA's Certificate
+openssl x509 -req -days 365 -in server_csr.pem -CA ca_cert.pem -CAkey ca_key.pem -CAcreateserial -out server_cert.pem
```
I think we should also mention how important is to set the CN field and why is it useful

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,52 @@ This example can be executed on any ESP32 board, the only required interface is
+
+Expiry time and metadata fields can be adjusted in the invocation.
+
+Please see the openssl man pages (man openssl-req) for more details.
```
```suggestion
Please see the openssl man pages (man openssl) for more details.
```

david-cermak commented on components/esp_websocket_client/examples/target/main/CMakeLists.txt:
```diff
@@ -1,2 +1,5 @@
 idf_component_register(SRCS "websocket_example.c"
-                    INCLUDE_DIRS ".")
+                    INCLUDE_DIRS "."
+                    EMBED_TXTFILES "certs/client_cert.pem"
```
Are we embedding these file unconditionally? Even if they are n ot used unde the default config?

david-cermak commented on components/esp_websocket_client/examples/target/main/Kconfig.projbuild:
```diff
@@ -20,6 +20,20 @@ menu "Example Configuration"
         help
             URL of websocket endpoint this example connects to and sends echo
 
+    config WS_OVER_TLS_SERVER_AUTH
```
Could you please check if `echo.websocket.events` provides a wss option? (would `wss://echo.websocket.events` work?)

david-cermak replied:
If we find a good public endpoint which supports wws, we don't have to use self-signed certs. And it would be more straight forward demonstration

david-cermak commented on components/esp_websocket_client/examples/target/main/certs/server_csr.pem:
we don't need this file

david-cermak commented on components/esp_websocket_client/examples/target/main/websocket_example.c:
```diff
@@ -143,6 +143,20 @@ static void websocket_app_start(void)
+    extern const char key_end[]   asm("_binary_client_key_pem_end");
+
+    // Please note if you don't use server over TLS scheme (wss), the certificates are not required
+    websocket_cfg.cert_pem = cacert_start;
+    websocket_cfg.client_cert = cert_start;
+    websocket_cfg.client_cert_len = cert_end - cert_start;
+    websocket_cfg.client_key = key_start;
+    websocket_cfg.client_key_len = key_end - key_start;
```
should we unconditionally configure all keys and certs?

david-cermak commented on components/esp_websocket_client/examples/target/pytest_websocket.py:
```diff
@@ -44,12 +47,25 @@ def send_data(self, data):
+    def __init__(self, port, use_tls, verify):
         self.port = port
+        self.use_tls = use_tls
+        self.serv_verify = verify
```
```suggestion
        self.client_verify = verify
```

david-cermak commented on components/esp_websocket_client/examples/target/main/certs/ca_key.pem:
Do we need this file?

gabsuren replied:
Good point. Removed

david-cermak commented on components/esp_websocket_client/examples/target/main/Kconfig.projbuild:
```diff
@@ -20,6 +20,26 @@ menu "Example Configuration"
 
+    config WS_OVER_TLS_SERVER_AUTH
+        bool "Enable WebSocket over TLS with Server Certificate Verification Only"
+        default n
```
Could we make this option enabled by default?

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
```diff
@@ -13,6 +13,53 @@ This example can be executed on any ESP32 board, the only required interface is
+* Example below demonstrates how to generate a new self signed certificates for the server and client using the OpenSSL command line tool
+
+### Generating a self signed Certificates with OpenSSL
+* The example below outlines the process for creating new certificates for both the server and client using OpenSSL, a widely-used command line tool for implementing TLS protocol:
```
Please rephrase this sentence, so it's clear that this example is a *very-very-very* simplified version of creating self signed certificates/keys with one common CA, no CN checks, not password protected, with hardcoded key sizes and types and it is used only for testing.

I see that you have such a note down below of that paragraph, but I think it's better to place it here.

david-cermak commented on components/esp_websocket_client/examples/target/README.md:
````diff
@@ -13,6 +13,55 @@ This example can be executed on any ESP32 board, the only required interface is
+
+```
+
+Expiry time and metadata fields can be adjusted in the invocation.
````
I can see that the expiration dates are set to a certain value, correct? like ~10years for the CA and 1 for others?

gabsuren replied:
updated to ~10 years, for others as well

david-cermak commented on components/esp_websocket_client/examples/target/sdkconfig.ci.tls_servert_auth:
Is this file needed if it's the default config?

gabsuren replied:
so you suggest to have 
`sdkconfig.ci.tls_mutual_auth` for mutual authentication 
`sdkconfig.ci.no_tls_auth` for none TLS authentication 
`sdkconfig.defaults` for tls server authentication ? 
Did I understood correctly ?

david-cermak replied:
I think we should just have
* the default config (maybe even without `sdkconfig.defaults` as everything is set to defaults)
* `sdkconfig.ci.mutual_auth` for mutual authentication

david-cermak reviewed (approved):
LGTM otherwise
